### PR TITLE
using URI query to communicate payload

### PR DESCRIPTION
**Pull request by makdharma**
When using GET verb, the request payload is now communicated as a base64
encoded string in the query parameter. Deleted the code that wrote/read
a custom header field (payload-bin), since caches don't consider this
field when looking for a cache hit.

### CODE REVIEW

File: src/core/lib/channel/http_client_filter.c [hunk 1]
```diff
@@ -292,16 +292,37 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
-                                          op->send_message->length));
-        error =
-            grpc_metadata_batch_add_tail(exec_ctx, op->send_initial_metadata,
-                                         &calld->payload_bin, payload_bin);
+        /* when all the send_message data is available, then modify the path
+         * MDELEM by appending base64 encoded query to the path */
+        static const int k_url_safe = 1;
+        static const int k_multi_line = 0;
+        static char *k_query_separator = "?";
+        char *strs_to_concatenate[3];
```
Comment (ctiller) [inline src/core/lib/channel/http_client_filter.c]: Use slice_malloc instead... will need to refactor b64 code to encode to a target buffer (and predict length)

File: src/core/lib/channel/http_server_filter.c [hunk 2]
```diff
@@ -196,6 +197,35 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
...
                                  GRPC_ERROR_STR_KEY, ":path"));
+  } else if (*calld->recv_cacheable_request == true) {
+    /* We have a cacheable request made with GET verb. The path contains the
+     * query parameter which is base64 encoded request payload. */
+    char *path =
+        grpc_dump_slice(GRPC_MDVALUE(b->idx.named.path->md), GPR_DUMP_ASCII);
+    static const char *QUERY_SEPARATOR = "?";
+    char **query_parts;
+    size_t num_query_parts;
+    gpr_string_split(path, QUERY_SEPARATOR, &query_parts, &num_query_parts);
```
Comment (ctiller) [inline src/core/lib/channel/http_server_filter.c]: slice_sub to avoid allocation

File: src/core/lib/channel/http_client_filter.c (hunk 1 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_server_filter.c (hunk 2 above)
Comment (makdharma) [inline src/core/lib/channel/http_server_filter.c]: done

**Comment by makdharma** [comment]
test this please

File: src/core/lib/channel/http_client_filter.c [hunk 3]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+         * hold combined path+query */
+        size_t estimated_len = GRPC_SLICE_LENGTH(path_slice);
+        estimated_len += strlen(k_query_separator);
+        estimated_len += grpc_base64_estimate_encoded_size(
+            op->send_message->length, k_url_safe, k_multi_line);
+        estimated_len += 1; /* for the trailing 0 */
+        grpc_slice path_with_query_slice = grpc_slice_malloc(estimated_len);
+
+        /* memcopy individual pieces into this slice */
+        char *write_ptr = (char *)GRPC_SLICE_START_PTR(path_with_query_slice);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: s/char/uint8_t

File: src/core/lib/channel/http_client_filter.c [hunk 4]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+
+        char *original_path = (char *)GRPC_SLICE_START_PTR(path_slice);
+        memcpy(write_ptr, original_path, GRPC_SLICE_LENGTH(path_slice));
+        write_ptr += (int)GRPC_SLICE_LENGTH(path_slice);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: don't cast to int. By using uint8_t instead of char for write_ptr, this won't complain about different signed types.

File: src/core/lib/channel/http_client_filter.c [hunk 5]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        estimated_len += strlen(k_query_separator);
+        estimated_len += grpc_base64_estimate_encoded_size(
+            op->send_message->length, k_url_safe, k_multi_line);
+        estimated_len += 1; /* for the trailing 0 */
+        grpc_slice path_with_query_slice = grpc_slice_malloc(estimated_len);
+
+        /* memcopy individual pieces into this slice */
+        char *write_ptr = (char *)GRPC_SLICE_START_PTR(path_with_query_slice);
+
+        char *original_path = (char *)GRPC_SLICE_START_PTR(path_slice);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: ditto

File: src/core/lib/channel/http_client_filter.c [hunk 6]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
-            grpc_slice_from_copied_buffer((const char *)calld->payload_bytes,
-                                          op->send_message->length));
-        error =
-            grpc_metadata_batch_add_tail(exec_ctx, op->send_initial_metadata,
-                                         &calld->payload_bin, payload_bin);
+        /* when all the send_message data is available, then modify the path
+         * MDELEM by appending base64 encoded query to the path */
+        static const int k_url_safe = 1;
+        static const int k_multi_line = 0;
+        static const char *k_query_separator = "?";
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]:
```
const char k_query_separator[] = "?";
const size_t k_query_separator_len = GPR_ARRAY_SIZE(k_query_separator);
```

File: src/core/lib/channel/http_client_filter.c [hunk 7]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
-        grpc_mdelem payload_bin = grpc_mdelem_from_slices(
-            exec_ctx, GRPC_MDSTR_GRPC_PAYLOAD_BIN,
-            grpc_slice_from_copied_buffer((const char *)calld->payload_bytes,
-                                          op->send_message->length));
-        error =
-            grpc_metadata_batch_add_tail(exec_ctx, op->send_initial_metadata,
-                                         &calld->payload_bin, payload_bin);
+        /* when all the send_message data is available, then modify the path
+         * MDELEM by appending base64 encoded query to the path */
+        static const int k_url_safe = 1;
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: why are these static? If their lifetime doesn't have to outlive this block, don't make them. That's memory used for the whole length of the process.

File: src/core/lib/channel/http_client_filter.c [hunk 8]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        static const int k_multi_line = 0;
+        static const char *k_query_separator = "?";
+
+        grpc_slice path_slice =
+            GRPC_MDVALUE(op->send_initial_metadata->idx.named.path->md);
+        /* sum up individual component's lengths and allocate enough memory to
+         * hold combined path+query */
+        size_t estimated_len = GRPC_SLICE_LENGTH(path_slice);
+        estimated_len += strlen(k_query_separator);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: use the `k_query_separator_len` suggested about instead of strlen all throughout.

File: src/core/lib/channel/http_client_filter.c [hunk 9]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+
+        /* memcopy individual pieces into this slice */
+        char *write_ptr = (char *)GRPC_SLICE_START_PTR(path_with_query_slice);
+
+        char *original_path = (char *)GRPC_SLICE_START_PTR(path_slice);
+        memcpy(write_ptr, original_path, GRPC_SLICE_LENGTH(path_slice));
+        write_ptr += (int)GRPC_SLICE_LENGTH(path_slice);
+
+        memcpy(write_ptr, k_query_separator, strlen(k_query_separator));
+        write_ptr += strlen(k_query_separator);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: k_query_separator_len

File: src/core/lib/channel/http_client_filter.c [hunk 10]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        grpc_slice path_with_query_slice = grpc_slice_malloc(estimated_len);
+
+        /* memcopy individual pieces into this slice */
+        char *write_ptr = (char *)GRPC_SLICE_START_PTR(path_with_query_slice);
+
+        char *original_path = (char *)GRPC_SLICE_START_PTR(path_slice);
+        memcpy(write_ptr, original_path, GRPC_SLICE_LENGTH(path_slice));
+        write_ptr += (int)GRPC_SLICE_LENGTH(path_slice);
+
+        memcpy(write_ptr, k_query_separator, strlen(k_query_separator));
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: k_query_separator_len

File: src/core/lib/channel/http_client_filter.c [hunk 11]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        write_ptr += strlen(k_query_separator);
+
+        grpc_base64_encode_core(write_ptr, calld->payload_bytes,
+                                op->send_message->length, k_url_safe,
+                                k_multi_line);
+
+        /* remove trailing unused memory and add trailing 0 to terminate string
+         */
+        char *t = (char *)GRPC_SLICE_START_PTR(path_with_query_slice);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: uint8_t

File: src/core/lib/channel/http_client_filter.c [hunk 12]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        size_t path_length = strlen(t) + 1;
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: GRPC_SLICE_LENGTH in lieu of strlen

File: src/core/lib/channel/http_client_filter.c [hunk 13]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        *(t + path_length) = 0;
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: nit: use '\0' instead of 0 for null chars

File: src/core/lib/channel/http_client_filter.c [hunk 14]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        path_with_query_slice =
+            grpc_slice_sub(path_with_query_slice, 0, path_length);
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: you may want to use `grpc_slice_sub_no_ref` instead

File: src/core/lib/channel/http_server_filter.c [hunk 15]
```diff
@@ -196,6 +197,38 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
     add_error(error_name, &error,
               grpc_error_set_str(GRPC_ERROR_CREATE("Missing header"),
                                  GRPC_ERROR_STR_KEY, ":path"));
+  } else if (*calld->recv_cacheable_request == true) {
+    /* We have a cacheable request made with GET verb. The path contains the
+     * query parameter which is base64 encoded request payload. */
+    static const char *k_query_separator = "?";
```
Comment (dgquintas) [inline src/core/lib/channel/http_server_filter.c]: same comments as above throughout this new block

File: src/core/lib/channel/http_client_filter.c (hunk 7 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 6 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 8 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 3 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 5 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 4 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 10 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 9 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 11 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 12 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: slice len won't match strlen (potentially, when b64encoder overestimated the length), hence this logic trims the slice length to match strlen.

File: src/core/lib/channel/http_client_filter.c (hunk 13 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: done

File: src/core/lib/channel/http_client_filter.c (hunk 14 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: Done

File: src/core/lib/channel/http_client_filter.c (hunk 12 above)
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: Consider using `strnlen` instead

File: src/core/lib/security/util/b64.h [hunk 16]
```diff
@@ -41,6 +41,17 @@
 char *grpc_base64_encode(const void *data, size_t data_size, int url_safe,
                          int multiline);
 
+/* estimate the upper bound on size of base64 encoded data. The actual size
+ * is guaranteed to be less than or equal to the size returned here. */
+size_t grpc_base64_estimate_encoded_size(size_t data_size, int url_safe,
```
Comment (dgquintas) [inline src/core/lib/security/util/b64.h]: s/int/bool

File: src/core/lib/security/util/b64.h [hunk 17]
```diff
@@ -41,6 +41,17 @@
...
+                                         int multiline);
+
+/* Encodes data using base64 and write it to memory pointed to by result. It is
+ * the caller's responsiblity to allocate enough memory in |result| to fit the
+ * encoded data. */
+void grpc_base64_encode_core(char *result, const void *vdata, size_t data_size,
+                             int url_safe, int multiline);
```
Comment (dgquintas) [inline src/core/lib/security/util/b64.h]: s/int/bool

File: src/core/lib/security/util/b64.h (hunk 16 above)
Comment (makdharma) [inline src/core/lib/security/util/b64.h]: I'd like to do this in a different PR so we keep the focus of this PR In just adding functionality.

File: src/core/lib/channel/http_client_filter.c [hunk 18]
```diff
@@ -297,7 +297,7 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
         const int k_url_safe = 1;
         const int k_multi_line = 0;
         const char *k_query_separator = "?";
-        const size_t k_query_separator_len = GPR_ARRAY_SIZE(k_query_separator);
+        const size_t k_query_separator_len = 1; /* strlen(k_query_separator) */
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: this is fragile. Why not GPR_ARRAY_SIZE?

File: src/core/lib/channel/http_server_filter.c [hunk 19]
```diff
@@ -214,7 +214,7 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
 
     /* substitute path metadata with just the path (not query) */
     grpc_mdelem mdelem_path_without_query = grpc_mdelem_from_slices(
-        exec_ctx, GRPC_MDSTR_PATH, grpc_slice_sub(path_slice, 0, offset));
+        exec_ctx, GRPC_MDSTR_PATH, grpc_slice_sub_no_ref(path_slice, 0, offset));
```
Comment (dgquintas) [inline src/core/lib/channel/http_server_filter.c]: nice! :)

File: src/core/lib/channel/http_client_filter.c (hunk 18 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: GPR_ARRAY_SIZE(x) != strlen(x). It crashes the test.

Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: GPR_ARRAY_SIZE(x) -1 == strlen(x) iff x is an array. That is, change k_query_separator from a pointer to `const char k_query_separator[] = "?"`

Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: The original comment has the whole suggestion: https://github.com/grpc/grpc/pull/10222#discussion_r107557882

File: src/core/lib/channel/http_server_filter.c [hunk 20]
```diff
@@ -209,26 +209,30 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
...
+      grpc_slice_buffer_add(
+          &calld->read_slice_buffer,
+          grpc_base64_decode(exec_ctx,
+                             (const char *)GRPC_SLICE_START_PTR(query_slice),
+                             k_url_safe));
+      grpc_slice_buffer_stream_init(&calld->read_stream,
+                                    &calld->read_slice_buffer, 0);
+      calld->seen_path_with_query = true;
+    } else {
+      gpr_log(GPR_ERROR, "GET request without QUERY");
```
Comment (dgquintas) [inline src/core/lib/channel/http_server_filter.c]: is it really an error for a GET to now have any query params? In any case it helps to print the offending query (although I wonder about security concerns...)

File: src/core/lib/channel/http_client_filter.c (hunk 12 above)
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: Ping?

File: src/core/lib/channel/http_client_filter.c [hunk 21]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
-                                          op->send_message->length));
-        error =
-            grpc_metadata_batch_add_tail(exec_ctx, op->send_initial_metadata,
-                                         &calld->payload_bin, payload_bin);
+        /* when all the send_message data is available, then modify the path
+         * MDELEM by appending base64 encoded query to the path */
+        const int k_url_safe = 1;
+        const int k_multi_line = 0;
+        const char *k_query_separator = "?";
+        const size_t k_query_separator_len = 1; /* strlen(k_query_separator) */
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: GPR_ARRAY_SIZE, after changing k_query_separator to be a char[] instead of a char*

File: src/core/lib/channel/http_client_filter.c [hunk 22]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
-        grpc_mdelem payload_bin = grpc_mdelem_from_slices(
-            exec_ctx, GRPC_MDSTR_GRPC_PAYLOAD_BIN,
-            grpc_slice_from_copied_buffer((const char *)calld->payload_bytes,
-                                          op->send_message->length));
-        error =
-            grpc_metadata_batch_add_tail(exec_ctx, op->send_initial_metadata,
-                                         &calld->payload_bin, payload_bin);
+        /* when all the send_message data is available, then modify the path
+         * MDELEM by appending base64 encoded query to the path */
+        const int k_url_safe = 1;
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: can you add a TODO to change these to `bool`?

File: src/core/lib/channel/http_client_filter.c [hunk 23]
```diff
@@ -292,19 +292,57 @@ static grpc_error *hc_mutate_op(grpc_exec_ctx *exec_ctx,
...
+        const int k_multi_line = 0;
+        const char *k_query_separator = "?";
```
Comment (dgquintas) [inline src/core/lib/channel/http_client_filter.c]: I'm not even sure we need a char*. Why not a single char? This is all RFC driven and that specifies it as a single char: https://tools.ietf.org/html/rfc3986#section-3.4

File: src/core/lib/channel/http_server_filter.c [hunk 24]
```diff
@@ -196,6 +197,42 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
     add_error(error_name, &error,
               grpc_error_set_str(GRPC_ERROR_CREATE("Missing header"),
                                  GRPC_ERROR_STR_KEY, ":path"));
+  } else if (*calld->recv_cacheable_request == true) {
+    /* We have a cacheable request made with GET verb. The path contains the
+     * query parameter which is base64 encoded request payload. */
+    const char *k_query_separator = "?";
```
Comment (dgquintas) [inline src/core/lib/channel/http_server_filter.c]: same comments as for the client version apply: why not a single char?

File: src/core/lib/channel/http_server_filter.c [hunk 25]
```diff
@@ -196,6 +197,42 @@ static grpc_error *server_filter_incoming_metadata(grpc_exec_ctx *exec_ctx,
...
+    grpc_slice path_slice = GRPC_MDVALUE(b->idx.named.path->md);
+    uint8_t *path_ptr = (uint8_t *)GRPC_SLICE_START_PTR(path_slice);
+    size_t path_length = GRPC_SLICE_LENGTH(path_slice);
+    /* offset of the character '?' */
+    size_t offset = 0;
+    for (offset = 0; *path_ptr != k_query_separator[0] && offset < path_length;
```
Comment (dgquintas) [inline src/core/lib/channel/http_server_filter.c]: using a single char would also do away with the need to do a [0]

File: src/core/lib/security/util/b64.c [hunk 26]
```diff
@@ -71,15 +71,31 @@ static const char base64_url_safe_chars[] =
...
-  char *result = gpr_malloc(result_projected_size);
+  return result_projected_size;
+}
+
+void grpc_base64_encode_core(char *result, const void *vdata, size_t data_size,
+                             int url_safe, int multiline) {
+  const unsigned char *data = vdata;
+  const char *base64_chars =
+      url_safe ? base64_url_safe_chars : base64_url_unsafe_chars;
+  size_t result_projected_size =
```
Comment (dgquintas) [inline src/core/lib/security/util/b64.c]: const

File: src/core/lib/channel/http_server_filter.c (hunk 20 above)
Comment (makdharma) [inline src/core/lib/channel/http_server_filter.c]: The only time we expect a GET req is when the payload is passed in as query. It is an error otherwise. Not sure if printing the path will  help debug.

File: src/core/lib/channel/http_client_filter.c (hunk 22 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: Added todo to b64.h which is used in multiple places.

File: src/core/lib/channel/http_client_filter.c (hunk 23 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: Good catch. replaced it with char in both filters.

File: src/core/lib/channel/http_client_filter.c (hunk 21 above)
Comment (makdharma) [inline src/core/lib/channel/http_client_filter.c]: not needed anymore.

File: src/core/lib/security/util/b64.c (hunk 26 above)
Comment (makdharma) [inline src/core/lib/security/util/b64.c]: Done.

**Comment by makdharma** [comment]
Known failures:
#10199 Windows arena_test.exe
#9551 ObjC Macos example build fail

Filed new issue:
#10317 Docker go image build failure (interop)
